src/main/scala/akka/contrib/process/NonBlockingProcess.scala

```diff
@@ -170,34 +170,51 @@ class NonBlockingProcess(
       override def onPreStart(nuProcess: NuProcess): Unit = {
         // Create our stream based actors away from this one given that we want them to continue
         // for a small while post the actor dying (it may take a tiny bit longer for the process
-        // to terminate).
-        implicit val stdioMat: ActorMaterializer = systemMat
+        // to terminate). Given the latent completion, we also ensure that the materializer is
+        // shutdown when we know we no longer require it. We use a reference counter for tracking
+        // usage. When either stdout or stderr are completed then we decrement the ref counter.
+        // When there are zero refs then we shutdown the materializer. Note that it is possible
+        // for a process to cause the exit handler to be invoked without having started (in the
+        // case of an error). We handle this scenario also.
+        val stdioMaterializer = ActorMaterializer()(context.system)
+        val stdioMaterializerRefCount = new AtomicInteger(2)
         val stdin =
           Sink
             .foreach[ByteString](bytes => nuProcess.writeStdin(bytes.toByteBuffer))
         val (out, stdout) =
           Source
             .queue[ByteString](stdoutBufferMaxChunks, OverflowStrategy.dropHead)
             .toMat(BroadcastHub.sink)(Keep.both)
-            .run
+            .run()(stdioMaterializer)
         val (err, stderr) =
           Source
             .queue[ByteString](stderrBufferMaxChunks, OverflowStrategy.dropHead)
             .toMat(BroadcastHub.sink)(Keep.both)
-            .run
+            .run()(stdioMaterializer)
+
+        def shutdownStdioMaterializer(): Unit =
+          synchronized(if (!stdioMaterializer.isShutdown) stdioMaterializer.shutdown())
 
         nuProcess.setProcessHandler(new NuAbstractProcessHandler {
           override def onStart(nuProcess: NuProcess): Unit =
             self ! Started(nuProcess.getPID, stdin, stdout, stderr)
 
-          override def onStderr(buffer: ByteBuffer, closed: Boolean): Unit =
+          override def onStderr(buffer: ByteBuffer, closed: Boolean): Unit = {
             enqueue(err, buffer, stderrMaxBytesPerChunk, closed)
+            if (closed)
+              if (stdioMaterializerRefCount.decrementAndGet() == 0) shutdownStdioMaterializer()
```

Comment:
I could have combined these two expressions, but given the side-effect of the `decrementAndGet` I felt this was clearer.